go-controller/pkg/metrics/metrics.go

```diff
@@ -28,6 +35,97 @@ var (
 	BuildDate string
 )
 
+type metricDetails struct {
+	help   string
+	metric prometheus.Gauge
+}
+
+// OVN/OVS components, namely ovn-northd, ovn-controller, and ovs-vswitchd provide various
+// metrics through the 'coverage/show' command. The following data structure holds all the
+// metrics we are interested in that output for a given component. We generalize capturing
+// these metrics across all OVN/OVS components.
+var componentCoverageShowMetricsMap = map[string]map[string]*metricDetails{}
+
+func parseMetricToFloat(componentName, metricName, value string) float64 {
+	f64Value, err := strconv.ParseFloat(value, 64)
+	if err != nil {
+		klog.Errorf("Failed to parse value %s into float for metric %s_%s :(%v)",
+			value, componentName, metricName, err)
+		return 0
+	}
+	return f64Value
+}
+
+// registerCoverageShowMetrics registers coverage/show metricss for
+// various components(ovn-northd, ovn-controller, and ovs-vswitchd) with prometheus
+func registerCoverageShowMetrics(target string, metricNamespace string, metricSubsystem string) {
+	coverageShowMetricsMap := componentCoverageShowMetricsMap[target]
+	for metricName, metricInfo := range coverageShowMetricsMap {
+		metricInfo.metric = prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Subsystem: metricSubsystem,
+			Name:      metricName,
+			Help:      metricInfo.help,
+		})
+		prometheus.MustRegister(metricInfo.metric)
+	}
+}
+
+// getCoverageShowOutputMap obtains the coverage/show metric values for the specified component.
+func getCoverageShowOutputMap(component string) (map[string]string, error) {
+	var stdout, stderr string
+	var err error
+
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("recovering from a panic while parsing the coverage/show output "+
+				"for %s: %v", component, r)
+		}
+	}()
+
+	if component == ovnNorthd {
+		stdout, stderr, err = util.RunOVNNorthAppCtl("coverage/show")
+	} else {
+		err = fmt.Errorf("component is unknown, and it isn't %s", ovnNorthd)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to get coverage/show output for %s "+
```

Comment:
if `component != ovnNorthd` should we return the previous err directly, otherwise this message will be misleading

Reply:
Agree.